pkg/util/map.go

```diff
@@ -0,0 +1,39 @@
+package util
+
+type GenerationalMap[K comparable, V any] struct {
+	oldgen map[K]V
+	newgen map[K]V
+
+	maxSize int
+	newV    func() V
+	gcCb    func()
+}
+
+// NewGenMap created which maintains at most maxSize recently used entries
+func NewGenMap[K comparable, V any](maxSize int, newV func() V, gcCb func()) GenerationalMap[K, V] {
```

Comment:
This would our first generic code in Loki I believe. Will wait for few more eyes from contributors.

Can we make it bit more readable may be? failing to understand what are `newV` and `gcCb` means here. Can we have may be better naming and some descriptive comment on those?

If we going with this generic map then, a have few tests for this map would be good.

Reply:
Just started to think why we need this Map in the first place? I see it's getting used only in one place and could be simple `map[string]*rate.Limiter`?. Trying to understand the rationale here.

Reply:
newV - new value
gcCb - garbage collector callback

I am happy to use any names you suggest for these or any other identifiers.

Purpose of this map is to keep maximum number of last used entries and GC the rest. Difference to LRU is that it doesn't pessimize common case of NOT reaching the limit.